e2e/internal/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,344 @@
+// TODO: //go:build e2eupgrade
+
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+package test
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/edgelesssys/constellation/v2/e2e/internal/kubectl"
+	"github.com/edgelesssys/constellation/v2/internal/attestation/measurements"
+	"github.com/edgelesssys/constellation/v2/internal/cloud/cloudprovider"
+	"github.com/edgelesssys/constellation/v2/internal/config"
+	"github.com/edgelesssys/constellation/v2/internal/constants"
+	"github.com/edgelesssys/constellation/v2/internal/file"
+	"github.com/edgelesssys/constellation/v2/internal/versionsapi"
+	"github.com/edgelesssys/constellation/v2/internal/versionsapi/fetcher"
+	"github.com/spf13/afero"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	coreV1 "k8s.io/api/core/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	toImage     = flag.String("to-image", "", "Image (shortversion) to upgrade to.")
+	toVersion   = flag.String("to-version", "", "CLI version to use.")
+	wantWorker  = flag.Int("want-worker", 0, "Number of wanted worker nodes.")
+	wantControl = flag.Int("want-control", 0, "Number of wanted control nodes.")
+)
+
+// E2E is creating cluster from root of Constellation repository, we need to
+// execute the upgrade test in that Constellation workspace.
+func TestMain(m *testing.M) {
+	if err := os.Chdir("../../.."); err != nil {
+		os.Exit(1)
+	}
+
+	os.Exit(m.Run())
+}
+
+func TestUpgrade(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+
+	k, err := kubectl.New()
+	require.NoError(err)
+	assert.NotNil(k)
+
+	testNodesEventuallyAvailable(t, k, *wantControl, *wantWorker)
+	testPodsEventuallyReady(t, k, "kube-system")
+
+	testCLIHasVersion(t, *toVersion)
+	wantImage := writeUpgradeConfig(t, *toImage)
+	cmd := exec.CommandContext(context.Background(), "constellation", "upgrade", "execute")
+	msg, err := cmd.CombinedOutput()
+	require.NoErrorf(err, "%s", string(msg))
+
+	testNodesEventuallyHaveImage(t, k, wantImage)
+}
+
+func writeUpgradeConfig(t *testing.T, toImage string) string {
```

Comment:
This would need some slight changes. We don't have the upgrade conf anymore since the `upgrade check`/`upgrade apply` refactoring that I merged Wednesday. Maybe we can do the required changes in a pair programming session later today?